include/sparrow/optional.hpp

```diff
@@ -0,0 +1,500 @@
+    
+    template <class T, class B, class U>
+    constexpr bool operator==(const optional<T, B>& lhs, const U& rhs) noexcept;
+
+    template <class T, class B, class U>
+    requires (!util::is_optional_v<U> && std::three_way_comparable_with<U, T>)
+    constexpr std::compare_three_way_result_t<T, U>
+    operator<=>(const optional<T, B>& lhs, const U& rhs) noexcept;
+
+    template <class T, class B, class U, class UB>
+    constexpr bool operator==(const optional<T, B>& lhs, const optional<U, UB>& rhs) noexcept;
+
+    template <class T, class B, std::three_way_comparable_with<T> U, class UB>
+    constexpr std::compare_three_way_result_t<T, U>
+    operator<=>(const optional<T, B>& lhs, const optional<U, UB>& rhs) noexcept;
+
+    template <class T, boolean_like B = bool>
+    constexpr optional<T, B> make_optional(T&& value, B&& flag = true);
+
+    /***************************
+     * optional implementation *
+     ***************************/
+
+    template <class T, boolean_like B>
+    constexpr bool optional<T, B>::has_value() const noexcept
+    {
+        return m_flag;
+    }
+
+    template <class T, boolean_like B>
+    constexpr optional<T, B>::operator bool() const noexcept
+    {
+        return m_flag;
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::operator*() & noexcept -> reference
+    {
+        return m_value;
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::operator*() const & noexcept -> const_reference
+    {
+        return m_value;
+    }
+    
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::operator*() && noexcept -> rvalue_reference
+    {
+        return std::move(m_value);
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::operator*() const && noexcept -> const_rvalue_reference
+    {
+        return std::move(m_value);
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::value() & -> reference
+    {
+        throw_if_empty();
+        return m_value;
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::value() const & -> const_reference
+    {
+        throw_if_empty();
+        return m_value;
+    }
+    
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::value() && -> rvalue_reference
+    {
+        throw_if_empty();
+        return std::move(m_value);
+    }
+
+    template <class T, boolean_like B>
+    constexpr auto optional<T, B>::value() const && -> const_rvalue_reference
+    {
+        throw_if_empty();
+        return std::move(m_value);
+    }
+
+    template <class T, boolean_like B>
+    template <class U>
+    constexpr auto optional<T, B>::value_or(U&& default_value) const & -> value_type
+    {
+        return bool(*this) ? **this : static_cast<value_type>(std::forward<U>(default_value)); 
```

Comment:
- `bool(*this)` the cast should not be necessary?
- `**this` I would prefer an explicit `value()` being called here.
- I would prefer the `static_cast` to be replaced by a construction (note that there is no concepts here like in the constructors, maybe there should be)

same issues below